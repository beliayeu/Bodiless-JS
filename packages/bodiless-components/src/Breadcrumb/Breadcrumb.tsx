/**
 * Copyright Â© 2020 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { ComponentType, HTMLProps } from 'react';
import { useNode } from '@bodiless/core';
import type { WithNodeKeyProps, WithNodeProps } from '@bodiless/core';
import { asComponent, designable, addProps } from '@bodiless/fclasses';
import type { DesignableComponentsProps } from '@bodiless/fclasses';
import { observer } from 'mobx-react-lite';
import { flowRight } from 'lodash';
import type { BreadcrumbItemType as BreadcrumbStoreItemType } from './BreadcrumbStore';
import { useBreadcrumbStore } from './BreadcrumbStoreProvider';

type BreadCrumbComponents = {
  StartingTrail: ComponentType<HTMLProps<HTMLSpanElement>>,
  Separator: ComponentType<HTMLProps<HTMLSpanElement>>,
  BreadcrumbWrapper: ComponentType<HTMLProps<HTMLUListElement>>,
  BreadcrumbItem: ComponentType<HTMLProps<HTMLLIElement>>,
  BreadcrumbLink: ComponentType<HTMLProps<HTMLAnchorElement> & WithNodeKeyProps>,
  BreadcrumbTitle: ComponentType<HTMLProps<HTMLSpanElement> & WithNodeKeyProps>,
  FinalTrail: ComponentType<HTMLProps<HTMLSpanElement>>,
};

type BreadcrumbItemKeys = {
  uuid: string | number;
  title: WithNodeProps;
  link: WithNodeProps;
};

type BreadcrumbProps = DesignableComponentsProps<BreadCrumbComponents> & {
  hasStartingTrail?: boolean | (() => boolean),
  items?: BreadcrumbItemKeys[],
  hasFinalTrail?: boolean | (() => boolean),
} & { };

const BreadcrumbClean$ = (props: BreadcrumbProps) => {
  const {
    hasStartingTrail = false,
    components,
    items = [],
    hasFinalTrail = false,
  } = props;
  const hasStartingTrail$ = typeof hasStartingTrail === 'function' ? hasStartingTrail() : hasStartingTrail;
  const hasFinalTrail$ = typeof hasFinalTrail === 'function' ? hasFinalTrail() : hasFinalTrail;
  const {
    StartingTrail,
    Separator,
    BreadcrumbWrapper,
    BreadcrumbItem,
    BreadcrumbLink,
    BreadcrumbTitle,
    FinalTrail,
  } = components;
  const items$ = items.map((item: BreadcrumbItemKeys, index: number) => {
    const isLastItem = index === (items.length - 1);
    return (
      <React.Fragment key={item.uuid}>
        <BreadcrumbItem key={item.uuid}>
          <BreadcrumbLink nodeKey={item.link.nodeKey} nodeCollection={item.link.nodeCollection}>
            <BreadcrumbTitle
              nodeKey={item.title.nodeKey}
              nodeCollection={item.title.nodeCollection}
            />
          </BreadcrumbLink>
        </BreadcrumbItem>
        {!isLastItem && <Separator key={`separator${item.uuid}`} />}
      </React.Fragment>
    );
  });
  return (
    <BreadcrumbWrapper>
      { hasStartingTrail$
        && (
        <>
          <BreadcrumbItem key="startingTrail">
            <StartingTrail />
          </BreadcrumbItem>
          { (items$.length > 0 || hasFinalTrail$)
            && <Separator key="startingTrailSeparator" />}
        </>
        )}
      {items$}
      { hasFinalTrail$
        && (
        <>
          { items$.length > 0
            && <Separator key="finalTrailSeparator" />}
          <BreadcrumbItem key="finalTrail">
            <FinalTrail />
          </BreadcrumbItem>
        </>
        )}
    </BreadcrumbWrapper>
  );
};

const BreadcrumbStartComponents: BreadCrumbComponents = {
  StartingTrail: asComponent('span'),
  Separator: asComponent('span'),
  BreadcrumbWrapper: asComponent('ul'),
  BreadcrumbItem: asComponent('li'),
  BreadcrumbLink: asComponent('a'),
  BreadcrumbTitle: asComponent('span'),
  FinalTrail: asComponent('span'),
};

const BreadcrumbClean = designable(BreadcrumbStartComponents)(BreadcrumbClean$);

// eslint-disable-next-line max-len
const withBreadcrumbItemsFromStore = (Component: ComponentType<BreadcrumbProps & WithNodeProps>) => {
  const WithBreadcrumbItemsFromStore = (props: BreadcrumbProps & WithNodeProps) => {
    const { nodeCollection, hasFinalTrail = false, ...rest } = props;
    const store = useBreadcrumbStore();
    if (store === undefined) return <Component {...props} />;
    const { node } = useNode(nodeCollection);
    const basePath = node.path;
    const items = store.breadcrumbTrail.map((item: BreadcrumbStoreItemType) => {
      const linkNodePath = item.link.nodePath.replace(`${basePath}$`, '');
      const titleNodePath = item.title.nodePath.replace(`${basePath}$`, '');
      return {
        uuid: item.uuid,
        link: {
          nodeKey: linkNodePath,
          nodeCollection,
        },
        title: {
          nodeKey: titleNodePath,
          nodeCollection,
        },
      };
    });
    const hasFinalTrail$0 = typeof hasFinalTrail === 'function' ? hasFinalTrail() : hasFinalTrail;
    const hasFinalTrail$1 = hasFinalTrail$0 && !store.hasLastItem();
    return <Component {...rest} items={items} hasFinalTrail={hasFinalTrail$1} />;
  };
  return WithBreadcrumbItemsFromStore;
};

const withStartingTrail = addProps({
  hasStartingTrail: true,
});

const withoutStartingTrail = addProps({
  hasStartingTrail: false,
});

const withFinalTrail = addProps({
  hasFinalTrail: true,
});

const withoutFinalTrail = addProps({
  hasFinalTrail: false,
});

const Breadcrumb = flowRight(
  observer,
  withBreadcrumbItemsFromStore,
)(BreadcrumbClean);

export {
  BreadcrumbClean,
  Breadcrumb,
  withStartingTrail as withBreadcrumbStartingTrail,
  withoutStartingTrail as withoutBreadcrumbStartingTrail,
  withFinalTrail as withBreadcrumbFinalTrail,
  withoutFinalTrail as withoutBreadcrumbFinalTrail,
};
