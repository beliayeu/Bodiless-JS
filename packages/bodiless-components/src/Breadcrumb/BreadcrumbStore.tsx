/**
 * Copyright Â© 2020 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { ComponentType } from 'react';
import { observable, action, computed } from 'mobx';

type BreadcrumbItemTitle = {
  // ToDo: Remove it, there is no need to store it here
  data: string | object;
  nodePath: string;
};
type BreadcrumbItemLink = {
  data: string ;
  nodePath: string;
};
type BreadcrumbItemSettings = {
  uuid: string;
  parent?: BreadcrumbItemInterface;
  title: BreadcrumbItemTitle;
  link: BreadcrumbItemLink;
  store: BreadcrumbStoreInterface;
};
export interface BreadcrumbItemInterface {
  setTitle: (title: BreadcrumbItemTitle) => void,
  setLink: (link: BreadcrumbItemLink) => void,
  getUUID: () => string,
  getTitle: () => BreadcrumbItemTitle;
  getLink: () => BreadcrumbItemLink;
  isSubpathOf: (item: BreadcrumbItemInterface | string) => boolean;
  hasPath: (item: BreadcrumbItemInterface | string) => boolean;
  isAncestorOf: (item: BreadcrumbItemInterface) => boolean;
  isDescendantOf: (item: BreadcrumbItemInterface) => boolean;
  getAncestors: () => BreadcrumbItemInterface[];
  getParent: () => BreadcrumbItemInterface | undefined;
};

const DEFAULT_URL_BASE = 'http://host';

const withoutTrailingSlash = (str: string) => str.replace(/\/$/, '');

const isChildOf = (child: string, parent: string) => {
  if (child === parent) return false;
  const childTokens = child.split('/').filter(i => i.length);
  const parentTokens = parent.split('/').filter(i => i.length);
  return parentTokens.every((t, i) => childTokens[i] === t);
}

export class BreadcrumbItem {
  uuid: string;
  parent?: BreadcrumbItemInterface;
  title: BreadcrumbItemTitle;
  link: BreadcrumbItemLink;
  store: BreadcrumbStoreInterface;

  constructor({
    uuid,
    title,
    link,
    store,
    parent
  }: BreadcrumbItemSettings) {
    this.uuid = uuid;
    this.parent = parent;
    this.title = title;
    this.link = link;
    this.store = store;
  }

  isSubpathOf(item: BreadcrumbItemInterface | string) {
    const base = typeof window === 'undefined'
      ? DEFAULT_URL_BASE
      : `${window.location.protocol}//${window.location.host}`;
    const itemURL = typeof item === 'string' ? new URL(item, base) : new URL(item.getLink().data, base);
    const thisURL = new URL(this.getLink().data, base);
    if (itemURL.host !== thisURL.host) return false;
    return isChildOf(itemURL.pathname, thisURL.pathname);
  }

  hasPath(item: BreadcrumbItemInterface | string) {
    const base = typeof window === 'undefined'
      ? DEFAULT_URL_BASE
      : `${window.location.protocol}//${window.location.host}`;
    const itemURL = typeof item === 'string' ? new URL(item, base) : new URL(item.getLink().data, base);
    const thisURL = new URL(this.getLink().data, base);
    if (itemURL.host !== thisURL.host) return false;
    return withoutTrailingSlash(thisURL.pathname) === withoutTrailingSlash(itemURL.pathname);
  }

  isDescendantOf(item: BreadcrumbItemInterface) {
    for (let current: BreadcrumbItemInterface | undefined = this.parent;
      current;
      current = this.parent
    ) if (current === item) return true;
    return false;
  }

  isAncestorOf(item: BreadcrumbItemInterface) {
    const isDescendant = item.isDescendantOf(this);
    return isDescendant;
  }

  getAncestors() {
    const ancestors = [];
    for (let current = this.getParent();
      current;
      current = current.getParent()
    ) {
      ancestors.push(current);
    }
    return ancestors;
  }

  getUUID() {
    return this.uuid;
  }

  getTitle() {
    return this.title;
  }

  getLink() {
    return this.link;
  }

  getParent() {
    return this.parent;
  }

  setTitle(title: BreadcrumbItemTitle) {
    this.title = title;
  }

  setLink(link: BreadcrumbItemLink) {
    this.link = link;
  }
};

interface BreadcrumbStoreInterface {
  setItem: (item: BreadcrumbItemInterface) => BreadcrumbItemInterface | undefined;
  deleteItem: (item: BreadcrumbItemInterface) => boolean;
  getPagePath: () => string;
  breadcrumbTrail: BreadcrumbItemInterface[];
  export: () => BreadcrumbItemInterface[];
  hasLastItem: () => boolean;
};

export class BreadcrumbStore implements BreadcrumbStoreInterface {
  @observable private items: Map<string, BreadcrumbItemInterface> = new Map<string, BreadcrumbItemInterface>();
  @observable private activeItem?: BreadcrumbItemInterface = undefined;
  private pagePath: string;

  constructor(pagePath: string ) {
    this.pagePath = pagePath;
  }

  @action private setActiveItem(item: BreadcrumbItemInterface) {
    console.log('setActiveItem into', item.getLink().data, item.getTitle().nodePath);
    this.activeItem = item;
  }

  @action setItem(item: BreadcrumbItemInterface) {
    if (
      item.isSubpathOf(this.pagePath) 
      && (!this.activeItem || this.activeItem.isSubpathOf(item))
    ) {
      this.setActiveItem(item);
    }
    this.items.set(item.getUUID(), item);
    return item;
  }

  deleteItem(item: BreadcrumbItemInterface | string) {
    const uuid = typeof item === 'string' ? item : item.getUUID();
    return this.items.delete(uuid);
  }

  getPagePath() {
    return this.pagePath;
  }

  @computed get breadcrumbTrail() {
    if (this.activeItem === undefined) return [];
    return [
      this.activeItem,
      ...this.activeItem.getAncestors(),
    ].reverse();
  }

  export() {
    return Array.from(this.items.values());
  }

  hasLastItem() {
    return this.activeItem !== undefined && this.activeItem.hasPath(this.pagePath);
  }
}

// ToDo: probably we should not have undefined
const BreadcrumbStoreContext = React.createContext<BreadcrumbStoreInterface | undefined>(undefined);
 
export const BreadcrumbStoreProvider: ComponentType<any> = ({ children, store }: any) => {
  return (
    <BreadcrumbStoreContext.Provider value={store}>{children}</BreadcrumbStoreContext.Provider>
  );
};
 
export const useBreadcrumbStore = () => React.useContext(BreadcrumbStoreContext);

// ToDo: refactor this since we know pagePath only in runtime
export const withBreadcrumbStore = (pagePath: string) => (Component: ComponentType) => (props: any) => {
  const store = new BreadcrumbStore(pagePath);
  return <Component {...props} store={store} />;
};