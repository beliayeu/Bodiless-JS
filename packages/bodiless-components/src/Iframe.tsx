/**
 * Copyright Â© 2020 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { ComponentType, HTMLProps } from 'react';

import {
  asBodilessComponent,
  ifEditable,
  useMenuOptionUI,
} from '@bodiless/core';
import type {
  AsBodiless,
  BodilessOptions,
} from '@bodiless/core';
import { addProps } from '@bodiless/fclasses';

import { flowRight } from 'lodash';
import withFormSnippet from './withFormSnippet';

// Type of the data used by this component.
export type Data = {
  src: string;
  height: string;
};

export type Props = HTMLProps<HTMLIFrameElement>;

// Options used to create an edit button.
const options: BodilessOptions<Props, Data> = {
  icon: 'settings',
  label: 'Settings',
  name: 'Edit',
  global: false,
  local: true,
  Wrapper: 'div',
  useCompoundForm: () => true,
};

const withoutPointerEvents = addProps({
  style: {
    pointerEvents: 'none',
  },
});

const withHeightSnippet = withFormSnippet({
  nodeKeys: 'height',
  defaultData: { height: '' },
  snippetOptions: {
    renderForm: () => {
      const { ComponentFormLabel, ComponentFormText } = useMenuOptionUI();
      return (
        <React.Fragment key="height">
          <ComponentFormLabel htmlFor="height">Height</ComponentFormLabel>
          <ComponentFormText field="height" />
        </React.Fragment>
      );
    },
  },
});

const withSrcSnippet = withFormSnippet({
  nodeKeys: 'src',
  defaultData: { src: '' },
  snippetOptions: {
    renderForm: () => {
      const { ComponentFormLabel, ComponentFormText } = useMenuOptionUI();
      return (
        <React.Fragment key="src">
          <ComponentFormLabel htmlFor="src">Src</ComponentFormLabel>
          <ComponentFormText field="src" />
        </React.Fragment>
      );
    },
  },
});

const asBaseBodilessIframe: AsBodiless<Props, Data> = (
  nodeKeys?,
  defaultData?,
  useOverrides?,
  Wrapper?: ComponentType<any> | string,
) => flowRight(
  ifEditable(withoutPointerEvents),
  asBodilessComponent({
    ...options,
    ...(Wrapper ? { Wrapper } : {}),
  })(nodeKeys, defaultData, useOverrides),
);

const asBodilessIframe: AsBodiless<Props, Data> = (
  nodeKeys?,
  defaultData?,
  useOverrides?,
  Wrapper?: ComponentType<any> | string,
) => flowRight(
  // @ts-ignore
  asBaseBodilessIframe(nodeKeys, defaultData, useOverrides, Wrapper),
  withHeightSnippet,
  withSrcSnippet,
);

export default asBodilessIframe;
export {
  asBaseBodilessIframe,
  withHeightSnippet,
  withSrcSnippet,
};
