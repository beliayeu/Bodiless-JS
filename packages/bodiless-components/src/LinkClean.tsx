/**
 * Copyright Â© 2020 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, {
  FC,
  ComponentType,
  Fragment,
  HTMLProps,
} from 'react';
import { flow } from 'lodash';
import { withNode } from '@bodiless/core';
import {
  StylableProps,
  A,
  DesignableComponentsProps,
  designable,
} from '@bodiless/fclasses';

export type LinkComponents = {
  Link: ComponentType<StylableProps>,
  Content: ComponentType<StylableProps>,
};

const linkComponentStart: LinkComponents = {
  Link: A,
  Content: Fragment,
};

type Props = DesignableComponentsProps<LinkComponents> & HTMLProps<HTMLElement>;

const LinkBase: FC<Props> = ({ components, ...rest }) => {
  const {
    Link,
    Content,
  } = components;

  return (
    <Link {...rest}>
      <Content />
    </Link>
  );
};

const LinkClean = flow(
  designable(linkComponentStart),
  withNode,
)(LinkBase);

export default LinkClean;
