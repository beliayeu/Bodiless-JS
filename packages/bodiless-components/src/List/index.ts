import asBodilessList, {
  asSubList, withSimpleSubListDesign,
} from './asBodilessList';

import asChameleonSubList, {
  withSubLists,
  withSubListDesign,
  withEmptySubListMarkup,
} from './asChameleonSubList';

import { asTestableList } from './List';

export {
  asBodilessList,
  asSubList,
  asChameleonSubList,
  withSimpleSubListDesign,
  withSubLists,
  withSubListDesign,
  withEmptySubListMarkup,
  asTestableList,
};

export type {
  FinalProps as ListProps,
  ItemProps as ListItemProps,
  ListDesignableComponents,
} from './types';
