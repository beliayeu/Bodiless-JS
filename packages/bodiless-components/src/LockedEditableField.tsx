/**
 * Copyright Â© 2020 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { ComponentType, HTMLProps, useState } from "react";
import {
  asComponent,
  DesignableComponentsProps,
  designable,
  withDesign,
  addProps,
} from "@bodiless/fclasses";

type Components = {
  Wrapper: ComponentType<HTMLProps<HTMLDivElement>>,
  Input: ComponentType<HTMLProps<HTMLInputElement>>,
  Preview: ComponentType<HTMLProps<HTMLDivElement>>,
  EditLink: ComponentType<HTMLProps<HTMLAnchorElement>>,
  CancelLink: ComponentType<HTMLProps<HTMLAnchorElement>>,
};

const startComponents: Components = {
  Wrapper: asComponent('div'),
  Input: asComponent('input'),
  Preview: asComponent('div'),
  EditLink: asComponent('a'),
  CancelLink: asComponent('a'),
};

type EditableFieldProps = DesignableComponentsProps<Components> & {
  value: string
};

const BaseLockedEditableField = (props: EditableFieldProps) => {
  const { components, value } = props;
  const [ isEdit, setIsEdit ] = useState<Boolean>(false);
  const {
    Wrapper,
    Input,
    Preview,
    EditLink,
    CancelLink,
  } = components;
  if (isEdit) {
    return (
      <Wrapper>
        <Input defaultValue={value} />
        <CancelLink onClick={() => setIsEdit(false)} />
      </Wrapper>
    );
  }
  return (
    <Wrapper>
      <Preview>{value}</Preview>
      <EditLink onClick={() => setIsEdit(true)} />
    </Wrapper>
  )
}

/**
 * Adds data- identifiers to help select elements in automated tests.
 */
const asTestableLockedEditableField = withDesign({
  Wrapper: addProps({ 'data-field-element': 'wrapper' }),
  Input: addProps({ 'data-field-element': 'input' }),
  Preview: addProps({ 'data-field-element': 'preview' }),
  EditLink: addProps({ 'data-field-element': 'edit' }),
  CancelLink: addProps({ 'data-field-element': 'cancel' }),
});

const CleanLockedEditableField = designable<Components>(startComponents)(BaseLockedEditableField);

export default CleanLockedEditableField;
export { asTestableLockedEditableField }

