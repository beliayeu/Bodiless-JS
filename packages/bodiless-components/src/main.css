.table {
  display: table
}

.relative {
  position: relative
}

.resize {
  resize: both
}

@keyframes spin {
  to {
    transform: rotate(360deg)
  }
}

@keyframes ping {
  75%, 100% {
    transform: scale(2);
    opacity: 0
  }
}

@keyframes pulse {
  50% {
    opacity: .5
  }
}

@keyframes bounce {
  0%, 100% {
    transform: translateY(-25%);
    animation-timing-function: cubic-bezier(0.8,0,1,1)
  }

  50% {
    transform: none;
    animation-timing-function: cubic-bezier(0,0,0.2,1)
  }
}

@media (min-width: 640px) {
}

@media (min-width: 768px) {
}

@media (min-width: 1024px) {
}

@media (min-width: 1280px) {
}