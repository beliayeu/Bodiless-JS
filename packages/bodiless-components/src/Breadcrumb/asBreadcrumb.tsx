/**
 * Copyright Â© 2020 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, {
  createContext, useContext, ComponentType, useRef, useEffect,
} from 'react';
import { useNode } from '@bodiless/core';
import { v4 } from 'uuid';
import { observer } from 'mobx-react-lite';
import { BreadcrumbItem } from './BreadcrumbStore';
import type { BreadcrumbItemType } from './BreadcrumbStore';
import { useBreadcrumbStore } from './BreadcrumbStoreProvider';

type LinkData = {
  href: string,
};

const breadcrumbContext = createContext<BreadcrumbItemType | undefined>(undefined);

export const useBreadcrumbContext = () => useContext(breadcrumbContext);
export const BreadcrumbContextProvider = breadcrumbContext.Provider;

export type BreadcrumbSettings = {
  linkNodeKey: string,
  titleNodeKey: string,
};

/**
 * Creates an HOC which specifies that a wrapped component is a breadcrumb. The HOC
 * will compare link data (href) from the specified nodekey with the current page
 * path and mark the breadcrumb as "active" if the current page matches or is a
 * subpage. When a breadcrumb is active, all parent breadcrumbs are also marked
 * as active.  The state of the breadcrumb can then be tested with the
 * `useIsBreadcrumbActive` hook, or with the `ifActiveVBreadcrumb` or
 * `ifNotActiveBreadcrumb` flow toggles, to control rendering.
 *
 * @param nodeKeys The nodekeys defining where to locate the link data defining this breadcrumb.
 *
 * @return An HOC which defines the wrapped component as a breadcrumb.
 */
const asBreadcrumb = ({
  linkNodeKey,
  titleNodeKey,
}: BreadcrumbSettings) => <P extends object>(Component: ComponentType<P>) => {
  const AsBreadcrumb = observer((props: P) => {
    const { node } = useNode();
    const titleNode = node.child<object>(titleNodeKey);
    const linkNode = node.child<LinkData>(linkNodeKey);
    const contextUuidRef = useRef(v4());
    const store = useBreadcrumbStore();
    if (store === undefined) return <Component {...props} />;
    const current = useBreadcrumbContext();
    const item = new BreadcrumbItem({
      uuid: contextUuidRef.current,
      title: {
        data: titleNode.data,
        nodePath: [...node.path, titleNodeKey].join('$'),
      },
      link: {
        data: linkNode.data.href,
        nodePath: [...node.path, linkNodeKey].join('$'),
      },
      parent: current,
      store,
    });
    store.setItem(item);
    // deleting item from store on unmount
    useEffect(() => () => {
      store.deleteItem(contextUuidRef.current);
    }, []);
    return (
      <BreadcrumbContextProvider value={item}>
        <Component {...props} />
      </BreadcrumbContextProvider>
    );
  });
  return AsBreadcrumb;
};

export default asBreadcrumb;
