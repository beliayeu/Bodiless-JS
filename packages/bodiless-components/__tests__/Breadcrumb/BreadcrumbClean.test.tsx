/**
 * Copyright Â© 2020 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { ComponentType } from 'react';
import { mount } from 'enzyme';
import {
  withDesign, asComponent, addProps, replaceWith, withoutProps,
} from '@bodiless/fclasses';
import { flowRight } from 'lodash';
import { BreadcrumbClean } from '../../src/Breadcrumb';

const withAttrRename = (oldAttr: string, newAttr: string) => (Component: ComponentType) => {
  const WithAttrRename = (props: any) => {
    const { [oldAttr]: oldAttrVal, ...rest } = props;
    const props$ = {
      ...rest,
      [newAttr]: oldAttrVal,
    };
    return <Component {...props$} />;
  };
  return WithAttrRename;
};

describe('BreadcrumbClean', () => {
  it('renders as empty unordered list by default', () => {
    // @ts-ignore
    const wrapper = mount(<BreadcrumbClean />);
    expect(wrapper.html()).toMatchSnapshot();
  });
  it('allows adding starting trail using design api', () => {
    const Breadcrumb = withDesign({
      StartingTrail: flowRight(
        addProps({
          className: 'starting-trail',
        }),
        replaceWith(asComponent('span')),
      ),
    // @ts-ignore
    })(BreadcrumbClean);
    // @ts-ignore
    const wrapper = mount(<Breadcrumb />);
    expect(wrapper.html()).toMatchSnapshot();
  });
  it('allows passing items as a prop', () => {
    const items = [
      {
        uuid: Math.random(),
        link: {
          nodeKey: 'linkNodeKey1',
        },
        title: {
          nodeKey: 'titleNodeKey1',
        },
      },
      {
        uuid: Math.random(),
        link: {
          nodeKey: 'linkNodeKey1',
        },
        title: {
          nodeKey: 'titleNodeKey1',
        },
      },
    ];
    const Breadcrumb = withDesign({
      BreadcrumbLink: flowRight(
        withAttrRename('nodeKey', 'href'),
        withoutProps('nodeCollection'),
      ),
      BreadcrumbTitle: flowRight(
        withAttrRename('nodeKey', 'aria-label'),
        withoutProps('nodeCollection'),
      ),
    // @ts-ignore
    })(BreadcrumbClean);
    // @ts-ignore
    const wrapper = mount(<Breadcrumb items={items} />);
    expect(wrapper.html()).toMatchSnapshot();
  });
  it('allows designing separator using design api', () => {
    const items = [
      {
        uuid: Math.random(),
        link: {
          nodeKey: 'linkNodeKey1',
        },
        title: {
          nodeKey: 'titleNodeKey1',
        },
      },
      {
        uuid: Math.random(),
        link: {
          nodeKey: 'linkNodeKey1',
        },
        title: {
          nodeKey: 'titleNodeKey1',
        },
      },
    ];
    const Breadcrumb = withDesign({
      BreadcrumbLink: flowRight(
        withAttrRename('nodeKey', 'href'),
        withoutProps('nodeCollection'),
      ),
      BreadcrumbTitle: flowRight(
        withAttrRename('nodeKey', 'aria-label'),
        withoutProps('nodeCollection'),
      ),
      Separator: addProps({
        className: 'separator',
      }),
    // @ts-ignore
    })(BreadcrumbClean);
    // @ts-ignore
    const wrapper = mount(<Breadcrumb items={items} />);
    expect(wrapper.html()).toMatchSnapshot();
  });
});
