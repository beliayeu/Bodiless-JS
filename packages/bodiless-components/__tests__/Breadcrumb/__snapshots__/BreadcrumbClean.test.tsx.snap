// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BreadcrumbClean allows adding starting trail and design it using design api 1`] = `"<ul><li><span aria-label=\\"starting-trail\\"></span></li></ul>"`;

exports[`BreadcrumbClean allows designing separator using design api 1`] = `"<ul><li><a href=\\"linkNodeKey1\\"><span aria-label=\\"titleNodeKey1\\"></span></a></li><span class=\\"separator\\"></span><li><a href=\\"linkNodeKey1\\"><span aria-label=\\"titleNodeKey1\\"></span></a></li></ul>"`;

exports[`BreadcrumbClean allows passing items as a prop 1`] = `"<ul><li><a href=\\"linkNodeKey1\\"><span aria-label=\\"titleNodeKey1\\"></span></a></li><span></span><li><a href=\\"linkNodeKey1\\"><span aria-label=\\"titleNodeKey1\\"></span></a></li></ul>"`;

exports[`BreadcrumbClean does not render separator after last item when final trail is not set 1`] = `"<ul><li><a href=\\"linkNodeKey1\\"><span aria-label=\\"titleNodeKey1\\"></span></a></li></ul>"`;

exports[`BreadcrumbClean renders as empty unordered list by default 1`] = `"<ul></ul>"`;
